physical/raft/fsm.go

```diff
@@ -653,3 +702,149 @@ func protoConfigurationToRaftConfiguration(configuration *ConfigurationValue) (u
+	}
+
+	opChunks, err := f.f.List(f.ctx, prefix)
+	if err != nil {
+		return false, errwrap.Wrapf("error listing all chunks for op: {{err}}", err)
+	}
+
+	return uint32(len(opChunks)) == chunk.NumChunks, nil
+}
+
+func (f *FSMChunkStorage) FinalizeOp(opNum uint64) ([]*raftchunking.ChunkInfo, error) {
+	ret, err := f.chunksForOpNum(opNum)
+	if err != nil {
+		return nil, errwrap.Wrapf("error getting chunks for op keys: {{err}}", err)
+	}
+
+	prefix, _ := f.chunkPaths(&raftchunking.ChunkInfo{OpNum: opNum})
+	if err := f.f.DeletePrefix(f.ctx, prefix); err != nil {
+		return nil, errwrap.Wrapf("error deleting prefix after op finalization: {{err}}", err)
+	}
+
+	return ret, nil
+}
+
+func (f *FSMChunkStorage) chunksForOpNum(opNum uint64) ([]*raftchunking.ChunkInfo, error) {
+	prefix, _ := f.chunkPaths(&raftchunking.ChunkInfo{OpNum: opNum})
+
+	opChunkKeys, err := f.f.List(f.ctx, prefix)
+	if err != nil {
+		return nil, errwrap.Wrapf("error fetching op chunk keys: {{err}}", err)
+	}
+
+	if len(opChunkKeys) == 0 {
+		return nil, nil
+	}
+
+	// Get the total number from the first key
+	totalNum, err := strconv.ParseInt(strings.Split(opChunkKeys[0], "-")[0], 10, 64)
+	if err != nil {
+		return nil, errwrap.Wrapf("error detecting total number of chunks: {{err}}", err)
+	}
+
+	ret := make([]*raftchunking.ChunkInfo, totalNum)
+
+	for _, v := range opChunkKeys {
+		seqNum, err := strconv.ParseInt(strings.Split(v, "-")[1], 10, 64)
+		if err != nil {
+			return nil, errwrap.Wrapf("error converting seqnum to integer: {{err}}", err)
+		}
+
+		entry, err := f.f.Get(f.ctx, prefix+v)
+		if err != nil {
+			return nil, errwrap.Wrapf("error fetching chunkinfo: {{err}}", err)
+		}
+
+		var ci raftchunking.ChunkInfo
+		if err := jsonutil.DecodeJSON(entry.Value, &ci); err != nil {
+			return nil, errwrap.Wrapf("error decoding chunkinfo json: {{err}}", err)
+		}
+
+		ret[seqNum] = &ci
+	}
+
+	return ret, nil
+}
+
+func (f *FSMChunkStorage) GetChunks() (raftchunking.ChunkMap, error) {
+	opNums, err := f.f.List(f.ctx, chunkingPrefix)
+	if err != nil {
+		return nil, errwrap.Wrapf("error doing recursive list for chunk saving: {{err}}", err)
+	}
+
+	if len(opNums) == 0 {
+		return nil, nil
+	}
+
+	ret := make(raftchunking.ChunkMap)
+	for _, opNumStr := range opNums {
+		opNum, err := strconv.ParseInt(opNumStr, 10, 64)
+		if err != nil {
+			return nil, errwrap.Wrapf("error parsing op num during chunk saving: {{err}}", err)
+		}
+
+		opChunks, err := f.chunksForOpNum(uint64(opNum))
+		if err != nil {
+			return nil, errwrap.Wrapf("error getting chunks for op keys during chunk saving: {{err}}", err)
+		}
+
+		ret[uint64(opNum)] = opChunks
+	}
+
+	return ret, nil
+}
+
+func (f *FSMChunkStorage) RestoreChunks(chunks raftchunking.ChunkMap) error {
+	if err := f.f.DeletePrefix(f.ctx, chunkingPrefix); err != nil {
+		return errwrap.Wrapf("error deleting prefix for chunk restoration: {{err}}", err)
+	}
+	if chunks == nil || len(chunks) == 0 {
```

Comment:
```suggestion
	if len(chunks) == 0 {
```
`len(chunks)` should already handle the nil case.